hnn_core/extracellular.py

```diff
@@ -465,40 +467,33 @@ def _build(self, cvode=None, include_celltypes='all'):
             ``'Pyr'``. For basket cells, use ``'Basket'``. NB This argument is
             currently not exposed in the API.
         """
-        secs_on_rank = h.allsec()  # get all h.Sections known to this MPI rank
+        self.secs_on_rank = h.allsec()  # get all h.Sections known to this rank
         _validate_type(include_celltypes, str)
         _check_option('include_celltypes', include_celltypes, ['all', 'Pyr',
                                                                'Basket'])
         if include_celltypes.lower() != 'all':
-            secs_on_rank = [s for s in secs_on_rank if
-                            include_celltypes in s.name()]
+            self.secs_on_rank = [s for s in self.secs_on_rank if
+                                 include_celltypes in s.name()]
 
-        segment_counts = [sec.nseg for sec in secs_on_rank]
-        n_total_segments = np.sum(segment_counts)
+        segment_counts = [sec.nseg for sec in self.secs_on_rank]
+        self.n_total_segments = np.sum(segment_counts)
 
         # pointers assigned to _ref_i_membrane_ at each EACH internal segment
-        self._nrn_imem_ptrvec = h.PtrVector(n_total_segments)
+        self._nrn_imem_ptrvec = h.PtrVector(self.n_total_segments)
         # placeholder into which pointer values are read on each sim time step
-        self._nrn_imem_vec = h.Vector(n_total_segments)
+        self._nrn_imem_vec = h.Vector(self.n_total_segments)
 
-        ptr_idx = 0
-        for sec in secs_on_rank:
-            for seg in sec:  # section end points (0, 1) not included
-                # set Nth pointer to the net membrane current at this segment
-                self._nrn_imem_ptrvec.pset(
-                    ptr_idx, sec(seg.x)._ref_i_membrane_)
-                ptr_idx += 1
-        if ptr_idx != n_total_segments:
-            raise RuntimeError(f'Expected {n_total_segments} imem pointers, '
-                               f'got {ptr_idx}.')
+        self._nrn_imem_ptrvec.ptr_update_callback(
+            self._set_imem_pointers_to_ref_i_membrane_)
```

Comment:
I don't quite understand the fix. How does moving to a method "fixes" the issue?

Reply:
> I don't quite understand the fix.

Neither do I :) My best bet is that `Cvode` (a global variable/object/concept that  is only created once) mistakenly tries to access memory locations of pointers that were created by the previous call to `_build`. Note that all our objects should be effectively deleted when we overwrite `net`. So the pointer vector is gone, but memory locations still being accessed. I'm just riffing here...

The docs [here](https://nrn.readthedocs.io/en/latest/python/programming/pointers/ptrvector.html#PtrVector.ptr_update_callback) mention [`CVode.cache_efficient()`](https://nrn.readthedocs.io/en/latest/python/simctrl/cvode.html#CVode.cache_efficient), but we don't actually set it:

https://github.com/jonescompneurolab/hnn-core/blob/78ce7cf73607135475a19718c79fa41304aaff80/hnn_core/network_builder.py#L209

However, it may be implicitly set by some other (`ParallelContext`) switch that we're not even aware of...

Again, I don't understand this either. Nor do I understand why a simple (fast) test doesn't trigger the bug.

As a point of interest, I figured out how [NetPyNe](https://github.com/Neurosim-lab/netpyne/blob/a90d2ea8db0aba791f6fa454d218f954df537c35/netpyne/sim/run.py#L99) and BMTK store the extracellular recordings at each time point, though I think implementing it is going to be a large PR...

Reply:
Very mysterious. I just spent 1.5 hours trying to manually delete the references to the pointers but that doesn't seem to work or maybe I didn't delete the right pointer. Would have preferred that route as it's not very elegant to use object attributes to make things work. Anyhow, I've decided it's not worth the time for now ... let's move on :)